File: packages/sdk-python/knowhere_sdk/services/table_fill.py
"""
表格填充服务
"""

from typing import Dict, Any, Optional
from ..client import KnowhereClient
from ..types import TableFillJobCreate, TableFillJobResponse, TableFillJobStatus


class TableFillService:
    """表格填充服务"""
    
    def __init__(self, client: KnowhereClient):
        self.client = client
    
    async def create_job(
        self,
        file_url: str,
        webhook_url: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> TableFillJobResponse:
        """
        创建表格填充任务
        
        Args:
            file_url: 表格文件URL
            webhook_url: Webhook通知URL（可选）
            metadata: 额外元数据（可选）
            
        Returns:
            TableFillJobResponse: 任务创建响应
        """
        data = {
            "file_url": file_url,
            "webhook_url": webhook_url,
            "metadata": metadata or {}
        }
        
        response = await self.client._request(
            "POST",
            "/v1/table-fill/jobs",
            data=data
        )
        
        return TableFillJobResponse(**response.data)
    
    async def get_job_status(self, job_id: str) -> TableFillJobStatus:
        """
        获取任务状态
        
        Args:
            job_id: 任务ID
            
        Returns:
            TableFillJobStatus: 任务状态
        """
        response = await self.client._request(
            "GET",
            f"/v1/table-fill/jobs/{job_id}"
        )
        
        return TableFillJobStatus(**response.data)
    
    async def download_result(self, job_id: str) -> bytes:
        """
        下载结果文件
        
        Args:
            job_id: 任务ID
            
        Returns:
            bytes: 结果文件内容
        """
        response = await self.client._request(
            "GET",
            f"/v1/table-fill/jobs/{job_id}/download"
        )
        
        return response.data
    
    async def wait_for_completion(
        self,
        job_id: str,
        timeout: int = 3600,
        poll_interval: int = 5
    ) -> TableFillJobStatus:
        """
        等待任务完成
        
        Args:
            job_id: 任务ID
            timeout: 超时时间（秒）
            poll_interval: 轮询间隔（秒）
            
        Returns:
            TableFillJobStatus: 最终任务状态
        """
        import asyncio
        import time
        
        start_time = time.time()
        
        while time.time() - start_time < timeout:
            status = await self.get_job_status(job_id)
            
            if status.status in ['completed', 'failed']:
                return status
            
            await asyncio.sleep(poll_interval)
        
        raise TimeoutError(f"任务 {job_id} 在 {timeout} 秒内未完成")

File: packages/sdk-typescript/src/services/tableFill.ts
/**
 * 表格填充服务
 */

import type { 
  TableFillJobCreate, 
  TableFillJobResponse, 
  TableFillJobStatus 
} from '../types';

export class TableFillService {
  private client: any; // KnowhereClient

  constructor(client: any) {
    this.client = client;
  }

  /**
   * 创建表格填充任务
   */
  async createJob(params: TableFillJobCreate): Promise<TableFillJobResponse> {
    const response = await this.client.request('/v1/table-fill/jobs', {
      method: 'POST',
      body: JSON.stringify(params),
    });

    return response.data;
  }

  /**
   * 获取任务状态
   */
  async getJobStatus(jobId: string): Promise<TableFillJobStatus> {
    const response = await this.client.request(`/v1/table-fill/jobs/${jobId}`);
    return response.data;
  }

  /**
   * 下载结果文件
   */
  async downloadResult(jobId: string): Promise<Blob> {
    const response = await fetch(`${this.client.config.baseUrl}/v1/table-fill/jobs/${jobId}/download`, {
      headers: {
        'Authorization': `Bearer ${this.client.config.apiKey}`,
      },
    });

    if (!response.ok) {
      throw new Error(`下载失败: ${response.statusText}`);
    }

    return response.blob();
  }

  /**
   * 等待任务完成
   */
  async waitForCompletion(
    jobId: string,
    timeout: number = 3600,
    pollInterval: number = 5000
  ): Promise<TableFillJobStatus> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeout * 1000) {
      const status = await this.getJobStatus(jobId);

      if (['completed', 'failed'].includes(status.status)) {
        return status;
      }

      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }

    throw new Error(`任务 ${jobId} 在 ${timeout} 秒内未完成`);
  }
}
